pulsar/producer_partition.go

```diff
@@ -1407,31 +1410,39 @@ func (p *partitionProducer) releaseSemaphoreAndMem(size int64) {
 	p.client.memLimit.ReleaseMemory(size)
 }
 
-func (p *partitionProducer) canAddToQueue(sr *sendRequest, uncompressedPayloadSize int64) bool {
+func (p *partitionProducer) canAddToQueue(sr *sendRequest) bool {
 	if p.options.DisableBlockIfQueueFull {
 		if !p.publishSemaphore.TryAcquire() {
 			runCallback(sr.callback, nil, sr.msg, errSendQueueIsFull)
 			return false
 		}
-		if !p.client.memLimit.TryReserveMemory(uncompressedPayloadSize) {
+	} else {
+		if !p.publishSemaphore.Acquire(sr.ctx) {
+			runCallback(sr.callback, nil, sr.msg, errContextExpired)
+			return false
+		}
+	}
+	p.metrics.MessagesPending.Inc()
+	p.metrics.BytesPending.Add(float64(len(sr.msg.Payload)))
```

Comment:
I think we should consider the `msg.Schema` here when using the Schema.
And why we add the bytes pending here instead of `canReserveMem` ?

Reply:
Thanks for your careful review. L1425-L1426 is the original code and I think we can move it to `canReserveMem` to fix this metric.

https://github.com/apache/pulsar-client-go/blob/e45122c2defc5efd4efc493d0acef278a7ccfc01/pulsar/producer_partition.go#L1436-L1441